web/src/copy_and_paste.js

```diff
@@ -293,29 +293,138 @@ function get_end_tr_from_endc($endc) {
+                .replace(/\n/gm, "\n  "); // custom 2 space indent
+            let prefix = "* ";
+            const parent = node.parentNode;
+            if (parent.nodeName === "OL") {
+                const start = parent.getAttribute("start");
+                const index = Array.prototype.indexOf.call(parent.children, node);
+                prefix = (start ? Number(start) + index : index + 1) + ". ";
+            }
+            return prefix + content + (node.nextSibling && !/\n$/.test(content) ? "\n" : "");
+        },
+    });
+    turndownService.addRule("zulipCodeBlock", {
+        filter(node, options) {
+            return (
+                options.codeBlockStyle === "fenced" &&
+                node.nodeName === "CODE" &&
+                node.parentElement?.nodeName === "PRE"
+            );
+        },
+
+        replacement(content, node, options) {
+            const language = node.closest(".codehilite")?.dataset?.codeLanguage || "";
+
+            const fenceChar = options.fence.charAt(0);
+            let fenceSize = 3;
+            const fenceInCodeRegex = new RegExp("^" + fenceChar + "{3,}", "gm");
+
+            let match;
+            while ((match = fenceInCodeRegex.exec(content))) {
+                if (match[0].length >= fenceSize) {
+                    fenceSize = match[0].length + 1;
+                }
+            }
+
+            const fence = fenceChar.repeat(fenceSize);
+
+            return (
+                "\n\n" +
+                fence +
+                language +
+                "\n" +
+                content.replace(/\n$/, "") +
+                "\n" +
+                fence +
+                "\n\n"
+            );
+        },
+    });
+    turndownService.addRule("zulipImagePreview", {
         filter(node) {
+            // select image previews in Zulip messages
             return (
-                node.nodeName === "A" && node.href === node.innerHTML && node.href === node.title
+                node.classList.contains("message_inline_image") && node.firstChild.nodeName === "A"
             );
         },
-        replacement(content) {
-            return content;
+
+        replacement(content, node) {
+            const copied_html = new DOMParser().parseFromString(paste_html, "text/html");
+            if (
+                node.firstChild.ariaLabel &&
+                !copied_html
+                    .querySelector("a[href='" + node.firstChild.getAttribute("href") + "']")
+                    ?.parentNode?.classList.contains("message_inline_image")
+            ) {
+                // if their generating link is also copied
+                return "";
+            }
+            return image_to_zulip_markdown(content, node.firstChild);
+        },
+    });
+    turndownService.addRule("images", {
+        filter: "img",
+
+        replacement: image_to_zulip_markdown,
+    });
+    turndownService.addRule("math", {
```

Comment:
Let's add a comment here explaining that we don't have a way to get the original `LaTeX` back from math, either with Zulip HTML or in general, so we just drop it, as we're likely to get garbage if we don't do so.

(In the future, we could have a `data-original-latex` feature on Zulip HTML if we wanted to make that work).